src/backend/cuda/kernel/canny.hpp

```diff
@@ -0,0 +1,372 @@
+                         (b2 * strong.strides[2] + b3 * strong.strides[3]) + strong.strides[1] + 1;
+          T* oPtr = output.ptr +
+                         (b2 * output.strides[2] + b3 * output.strides[3]) + output.strides[1] + 1;
+
+    if (gx<(output.dims[0]-2) && gy<(output.dims[1]-2))
+    {
+        int idx   = lIdx(gx, gy, output.strides[0], output.strides[1]);
+        oPtr[idx] = (sPtr[idx] > 0 ? STRONG : (wPtr[idx] > 0 ? WEAK : NOEDGE));
+    }
+}
+
+// hasChanged is a variable in kernel space
+// used to track the convergence of
+// the breath first search algorithm
+__device__ int hasChanged = 0;
+
+#define VALID_BLOCK_IDX(j, i) ( (j)>0 && (j)<(SHRD_MEM_HEIGHT-1) && (i)>0 && (i)<(SHRD_MEM_WIDTH-1) )
+
+template<typename T>
+static __global__
+void edgeTrackKernel(Param<T> output, unsigned nBBS0, unsigned nBBS1)
+{
+    const unsigned SHRD_MEM_WIDTH  = THREADS_X + 2; // Cols
+    const unsigned SHRD_MEM_HEIGHT = THREADS_Y + 2; // Rows
+
+    // shared memory with 1 pixel border
+    // strong and weak images are binary(char) images thus,
+    // occupying only (16+2)*(16+2) = 324 bytes per shared memory tile
+    __shared__ int outMem  [ SHRD_MEM_HEIGHT ] [ SHRD_MEM_WIDTH ];
+    __shared__ int numChanges;
+
+    // local thread indices
+    const int lx = threadIdx.x;
+    const int ly = threadIdx.y;
+
+    // batch offsets for 3rd and 4th dimension
+    const unsigned b2 = blockIdx.x / nBBS0;
+    const unsigned b3 = blockIdx.y / nBBS1;
+
+    // global indices
+    const int gx = blockDim.x * (blockIdx.x-b2*nBBS0) + lx;
+    const int gy = blockDim.y * (blockIdx.y-b3*nBBS1) + ly;
+
+    // Offset input and output pointers to second pixel of second coloumn/row
+    // to skip the border
+    T* oPtr = output.ptr + (b2 * output.strides[2] + b3 * output.strides[3]) + output.strides[1] + 1;
+
+    // pull image to shared memory
+#pragma unroll
+    for (int b=ly, gy2=gy; b<SHRD_MEM_HEIGHT; b+=blockDim.y, gy2+=blockDim.y)
+#pragma unroll
+        for (int a=lx, gx2=gx; a<SHRD_MEM_WIDTH; a+=blockDim.x, gx2+=blockDim.x)
+            outMem[b][a]  = oPtr[ lIdx(gx2-1, gy2-1, output.strides[0], output.strides[1]) ];
+
+    int i = lx + 1;
+    int j = ly + 1;
+
+    if (lx==0 && ly==0)
+        numChanges = 0;
+
+    __syncthreads();
+
+    int continueIter = 1;
+
+    while (continueIter)
+    {
+        int cu = outMem[j][i];
+        int nw = outMem[j-1][i-1];
+        int no = outMem[j-1][i  ];
+        int ne = outMem[j-1][i+1];
+        int ea = outMem[j  ][i+1];
+        int se = outMem[j+1][i+1];
+        int so = outMem[j+1][i  ];
+        int sw = outMem[j+1][i-1];
+        int we = outMem[j  ][i-1];
+
+        bool hasStrongNeighbour = nw==STRONG || no==STRONG || ne==STRONG || ea==STRONG ||
+                                  se==STRONG || so==STRONG || sw==STRONG || we==STRONG;
+
+        if (cu==WEAK && hasStrongNeighbour)
+        {
+            outMem[j][i] = STRONG;
+            atomicAdd(&numChanges, 1);
```

Comment:
May be I wasn't clear, my question is about reading the value of `hasChanged`.  Isn't that global variable and hence result in global memory(which has higher latency) read ? 

Since it is single variable, do you think it will be a broadcast ?